### ISSUE TITLE
Improve startup performance

### ISSUE THREAD

**Issue by jamonholmgren** [infinitered/gluegun#355]
Gluegun currently takes about 480ms to boot. That's compared to about 115ms for `node --version` on my machine.

I think we could improve startup performance by about 30-35% by disabling unneeded core extensions. The three that cause the most performance penalty are:

```
prompt +100ms, +4kb
print +45ms, +8kb
http +30ms, +0kb
system +10ms, +8kb
```

The rest are negligeable.

### API changes

The builder API would be something like this:

```js
  const cli = build()
    .brand('gluegun')
    .exclude(['semver', 'system', 'prompt', 'http'])
    .src(__dirname)
    .help()
    .version()
    .create()
```

I created a WIP branch and this cut Gluegun startup time from 480ms to 315ms, which is nontrivial. Removing the `.exclude()` call brought it right back up to 480ms, so it's working as intended. Disabling _all_ core extensions brought it down to 275ms, but that's mostly the `print` core extensions which is almost always needed.

### Downsides

1. We won't be able to use Rollup anymore. This is because Rollup needs to know whether to include the core extensions in the bundle or not. I tried a few things and concluded the best thing to do would be to go away from Rollup and ship an NPM package with all the files instead of just a rolled-up index.js
2. We won't be able to export the core extensions from gluegun proper anymore, for example, `import { prompt } from 'gluegun'` wouldn't work. To be honest, this has never felt very clean to me anyway, so it doesn't sadden me to see it go.

### WIP branch

Go here to see my WIP: https://github.com/infinitered/gluegun/compare/perf/startup-time

I'd like to hear thoughts about this, ideas, anything else.

**Comment by skellock** [comment]
Ditching #2 is really bad news for me.  I use that everywhere.  

I disagree with the clean statement. The idiomatic way to import things in `js` and `ts` is via `import/require`.

Gluegun has two different ways to deliver functions to you, with & without context.  Perhaps this is the thing that feels a bit unclean to you (the fact there's two ways)?

Without direct exports, you're left shoehorning `Context` into every function call when all you need is a handful of `export { strings } from 'gluegun'`.  It gets ugly quick.

So perhaps some discussion around what `Context` should mean might be a good start?

It currently means: Tools + Context of the current user's request.

**Comment by jamonholmgren** [comment]
Good feedback! You’re the biggest user of gluegun and it’s good to hear the use cases against this optimization. I’m more sad about losing Rollup, so if you don’t like losing #2 and I don’t like losing #1 then that’s a pretty strong argument against going down this path.

It’s really good to get some perf numbers though. Perhaps we could find ways to fix the prompt and print loading times without this sort of optimization.

One thought - we could provide a separate object out of gluegun where you could retrieve tools on demand. Something like:

```js
import { build, createToolbox } from “gluegun”

const { print, filesystem } = createToolbox()
```

(Not in love with the `createToolbox` name, but you get the idea)

**Comment by jamonholmgren** [comment]
Maybe `createToolbox` could even take arguments to get specific tools, if you want.

**Comment by skellock** [comment]
Well, I need them, but I don’t need them bootstrapped.  For example, if we moved them to `gluegun/strings`, etc, that wouldn’t be a huge refactor for me at all.

I wonder if we could have the old behaviour hanging off `gluegun/sloth`? That way we would opt in to slowness.

I wouldn’t want you to stop these efforts!

**Comment by jamonholmgren** [comment]
That's an interesting idea! I'll play with that. Perhaps `gluegun/toolbox`, to stay with the theme.

**Comment by jamonholmgren** [comment]
Closed in #357.
